src/main/java/org/spongepowered/api/util/annotation/DoNotStore.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.util.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Target;
+
+/**
+ * The instance of the class to which this annotation is applied cannot be stored.
+ *
+ * The annotation is applied to classes to indicate that their instances should not
+ * be stored in a field or top-level variable.
+ * The annotation is inherited.
+ */
+@Inherited
+@Target(ElementType.TYPE)
+public @interface DoNotStore {
```

Comment:
Add `@Documented` as this is a documentation marker so should appear in the javadocs and is considered part of the public contract.

I'd change the javadocs on this too to give developers a hint on what (not) to do, something like:

```
Objects of types that are annotated with this method <strong>must not</strong> be stored as they may be invalidated at any time.

<p>In particular, you should not:</p>

<ul>
    <li>Store annotated objects on a field</li>
    <li>Cache annotated objects in any collection</li>
</ul>

<p>Doing so may result in memory leaks and/or result in you operating on a stale object, meaning your operations will fail, often silently.</p>

<p>Where objects have identifiers (usually a {@link UUID} or name), you should store the identifiers instead, and use these to get the required object from the API when needed.</p>
```